### PR TITLE
[GlobalISel] Check width of APInts in Reassoc PtrAdd combine

### PR DESCRIPTION
**Pull request by mbrkusanin**

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -4653,7 +4653,8 @@ bool CombinerHelper::matchReassocFoldConstantsInSubTree(GPtrAdd &MI,
     return false;
 
   MatchInfo = [=, &MI](MachineIRBuilder &B) {
-    auto NewCst = B.buildConstant(MRI.getType(Src2Reg), *C1 + *C2);
+    auto NewCst = B.buildConstant(MRI.getType(Src2Reg),
+                                  C1->sextOrTrunc(C2->getBitWidth()) + *C2);
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4657]: `
unsigned BitWidth =
        std::max(C1->getBitWidth(), C2->getBitWidth());
`

see https://github.com/llvm/llvm-project/pull/82927

**Comment by jayfoad** [comment]
Allowing arbitrary widths for the RHS of G_PTR_ADD makes me very uncomfortable. I would much prefer enforcing that the RHS has the same bitwidth as the index size of the pointer's address space. See also #81473.

**Comment by tschuett** [comment]
If you ask for `getIConstantVRegValWithLookThrough` from a s64 register, there is no guarantee that the bitwidth of APInt is 64 bit. If I have some fancy non integral AS with 128bit pointers, why should the offset of `G_PTR_ADD` be 128bit?

The test case is using the pre-legalizer combiner. I bet the real assert was in a bigger input.

**Comment by jayfoad** [comment]
> If you ask for getIConstantVRegValWithLookThrough from a s64 register, there is no guarantee that the bitwidth of APInt is 64 bit.

Yes that should be guaranteed. If `getIConstantVRegValWithLookThrough` looks through a truncation or extension, it truncates or extends the APInt accordingly before returning it. Do you have an example where this does not work?

**Comment by jayfoad** [comment]
> Allowing arbitrary widths for the RHS of G_PTR_ADD makes me very uncomfortable. I would much prefer enforcing that the RHS has the same bitwidth as the index size of the pointer's address space. See also #81473.

#84352

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 2]
```diff
@@ -4652,14 +4652,16 @@ bool CombinerHelper::matchReassocFoldConstantsInSubTree(GPtrAdd &MI,
   if (!C2)
     return false;
 
-  // If constant widths differ pick larger one.
-  unsigned BitWidth = std::max(C1->getBitWidth(), C2->getBitWidth());
-  APInt NewConst = C1->sext(BitWidth) + C2->sext(BitWidth);
-  LLT Type =
-      MRI.getType(C1->getBitWidth() > C2->getBitWidth() ? LHSSrc2 : Src2Reg);
+  // Pick correct size for the constant based on the address space
```
Comment (jayfoad) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4655]: If #84352 is accepted then this won't be necessary - you can rely on C1 and C2 both having the correct (same) width already.

Comment (mbrkusanin) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4655]: Ok then

**Pull request closed without merging**